File: backend/app/schemas/real_world_safety.py
from typing import Literal

from pydantic import BaseModel, Field


RealWorldSafetySourceKind = Literal["structured_api", "public_notice", "normalized_public_notice"]


class RealWorldSafetyRecord(BaseModel):
    source_name: str
    source_type: str
    source_url: str
    source_kind: RealWorldSafetySourceKind
    category: str | None = None
    product_name: str | None = None
    brand_name: str | None = None
    company_name: str | None = None
    title: str | None = None
    reason: str | None = None
    hazard_type: str | None = None
    remedy: str | None = None
    published_date: str | None = None
    recall_number: str | None = None
    affected_models: list[str] = Field(default_factory=list)
    affected_lots: list[str] = Field(default_factory=list)
    raw_payload_hash: str
    retrieved_at: str
    record_url: str | None = None
    extraction_confidence: str | None = None
    source_text_excerpt: str | None = None


class RealWorldSafetyCheckedSource(BaseModel):
    source_id: str
    source_name: str
    source_type: str
    source_url: str
    source_kind: RealWorldSafetySourceKind
    upstream_status: str
    record_count: int


class RealWorldSafetyFailedSource(BaseModel):
    source_id: str
    source_name: str
    source_type: str
    source_url: str
    source_kind: RealWorldSafetySourceKind
    error_type: str
    reason: str


class RealWorldSafetyAuditSummary(BaseModel):
    audit_id: str
    source_id: str
    source_name: str
    module: str
    upstream_status: str
    record_count: int
    transform_version: str
    source_snapshot_status: str | None = None
    source_pull_id: str | None = None
    source_payload_hash: str | None = None


class RealWorldSafetySourceFreshness(BaseModel):
    source_id: str
    source_name: str
    source_type: str
    source_kind: RealWorldSafetySourceKind
    upstream_status: str
    record_count: int
    freshness_status: str
    user_label: str
    explanation: str
    source_snapshot_status: str | None = None
    source_pull_id: str | None = None
    source_payload_hash: str | None = None
    checked_at: str


class RealWorldSafetyQueryUnderstanding(BaseModel):
    original_query: str
    normalized_query: str
    search_query: str
    corrections_applied: list[str] = Field(default_factory=list)
    expanded_terms: list[str] = Field(default_factory=list)
    expansion_search_terms_used: list[str] = Field(default_factory=list)
    detected_identifiers: dict[str, str | None]
    query_type_hints: list[str] = Field(default_factory=list)


class RealWorldSafetySearchPlan(BaseModel):
    intent: str
    confidence: str
    reason: str
    primary_source_ids: list[str] = Field(default_factory=list)
    secondary_source_ids: list[str] = Field(default_factory=list)
    sources_to_check: list[str] = Field(default_factory=list)
    clarification_required: bool = False


class RealWorldSafetyIntelligenceSummary(BaseModel):
    query_type: str
    recall_or_enforcement_found: bool
    reference_or_label_found: bool
    signal_report_found: bool
    outbreak_context_found: bool = False
    matched_sources_by_role: dict[str, list[str]]
    checked_sources_by_role: dict[str, list[str]]
    top_result_titles: list[str] = Field(default_factory=list)
    expansion_explanations: list[str] = Field(default_factory=list)
    plain_language_summary: str
    suggested_next_steps: list[str] = Field(default_factory=list)
    caveat: str


class RealWorldSafetyIdentifierCheckItem(BaseModel):
    type: str
    label: str
    value: str | None = None
    source: str
    reason: str


class RealWorldSafetyIdentifierCheck(BaseModel):
    detected: list[RealWorldSafetyIdentifierCheckItem] = Field(default_factory=list)
    to_verify: list[RealWorldSafetyIdentifierCheckItem] = Field(default_factory=list)
    user_message: str


class RealWorldSafetySearchResponse(BaseModel):
    query: str
    raw_query: str
    query_understanding: RealWorldSafetyQueryUnderstanding
    search_plan: RealWorldSafetySearchPlan
    count: int
    limit: int
    retrieval_timestamp: str
    sources_checked: list[RealWorldSafetyCheckedSource]
    sources_failed: list[RealWorldSafetyFailedSource]
    records_per_source: dict[str, int]
    structured_api_matches: int
    public_notice_matches: int
    total_matches: int
    no_match_explanation: str | None = None
    safety_intelligence_summary: RealWorldSafetyIntelligenceSummary
    identifier_check: RealWorldSafetyIdentifierCheck
    public_data_disclaimer: str
    limitations: list[str]
    source_audits: list[RealWorldSafetyAuditSummary]
    source_freshness: list[RealWorldSafetySourceFreshness]
    results: list[RealWorldSafetyRecord]

File: frontend/src/api/realWorldSafety.ts
import { apiClient } from './client'

export type RealWorldSafetySourceKind = 'structured_api' | 'public_notice'
export type RealWorldSafetySort = 'score' | 'latest'

export type RealWorldSafetyRecord = {
  source_name: string
  source_type: string
  source_url: string
  source_kind: RealWorldSafetySourceKind
  category: string | null
  product_name: string | null
  brand_name: string | null
  company_name: string | null
  title: string | null
  reason: string | null
  hazard_type: string | null
  remedy: string | null
  published_date: string | null
  recall_number: string | null
  affected_models: string[]
  affected_lots: string[]
  raw_payload_hash: string
  retrieved_at: string
  record_url: string | null
}

export type RealWorldSafetyCheckedSource = {
  source_id: string
  source_name: string
  source_type: string
  source_url: string
  source_kind: RealWorldSafetySourceKind
  upstream_status: string
  record_count: number
}

export type RealWorldSafetyFailedSource = {
  source_id: string
  source_name: string
  source_type: string
  source_url: string
  source_kind: RealWorldSafetySourceKind
  error_type: string
  reason: string
}

export type RealWorldSafetyQueryUnderstanding = {
  original_query: string
  normalized_query: string
  search_query: string
  corrections_applied: string[]
  expanded_terms: string[]
  expansion_search_terms_used: string[]
  detected_identifiers: Record<'vin' | 'ndc' | 'upc', string | null>
  query_type_hints: string[]
}

export type RealWorldSafetySourceRole =
  | 'recall_enforcement'
  | 'reference_identity'
  | 'label_reference'
  | 'signal_report'
  | 'other'

export type RealWorldSafetyIntelligenceSummary = {
  query_type: string
  recall_or_enforcement_found: boolean
  reference_or_label_found: boolean
  signal_report_found: boolean
  matched_sources_by_role: Record<RealWorldSafetySourceRole, string[]>
  checked_sources_by_role: Record<RealWorldSafetySourceRole, string[]>
  top_result_titles: string[]
  expansion_explanations: string[]
  plain_language_summary: string
  suggested_next_steps: string[]
  caveat: string
}

export type RealWorldSafetyAuditSummary = {
  audit_id: string
  source_id: string
  source_name: string
  module: string
  upstream_status: string
  record_count: number
  transform_version: string
  source_snapshot_status: string | null
  source_pull_id: string | null
  source_payload_hash: string | null
}

export type RealWorldSafetySearchResponse = {
  query: string
  raw_query: string
  query_understanding: RealWorldSafetyQueryUnderstanding
  count: number
  limit: number
  retrieval_timestamp: string
  sources_checked: RealWorldSafetyCheckedSource[]
  sources_failed: RealWorldSafetyFailedSource[]
  records_per_source: Record<string, number>
  structured_api_matches: number
  public_notice_matches: number
  total_matches: number
  no_match_explanation: string | null
  safety_intelligence_summary: RealWorldSafetyIntelligenceSummary
  public_data_disclaimer: string
  limitations: string[]
  source_audits: RealWorldSafetyAuditSummary[]
  results: RealWorldSafetyRecord[]
}

export async function searchRealWorldSafety(
  query: string,
  limit = 10,
  sort: RealWorldSafetySort = 'score',
): Promise<RealWorldSafetySearchResponse> {
  const response = await apiClient.get<RealWorldSafetySearchResponse>(
    '/api/v1/real-world-safety/search',
    {
      params: {
        q: query,
        limit,
        sort,
      },
    },
  )

  return response.data
}
